ksp_plugin_adapter/manœuvre_marker.cs

```diff
@@ -0,0 +1,210 @@
+
+    get_burn_ = get_burn;
+    modify_burn_ = modify_burn;
+
+    transform.position = ScaledSpace.LocalToScaledSpace(world_position);
+
+    tangent_.transform.localRotation = UnityEngine.Quaternion.FromToRotation(
+        UnityEngine.Vector3.up,
+        (Vector3d)manœuvre.tangent);
+    normal_.transform.localRotation = UnityEngine.Quaternion.FromToRotation(
+        UnityEngine.Vector3.up,
+        (Vector3d)manœuvre.normal);
+    binormal_.transform.localRotation = UnityEngine.Quaternion.FromToRotation(
+        UnityEngine.Vector3.up,
+        (Vector3d)manœuvre.binormal);
+
+    normalized_scale_ =
+        (ScaledSpace.ScaledToLocalSpace(MapView.MapCamera.transform.position) -
+         world_position).magnitude *
+        0.015 *
+        ScaledSpace.InverseScaleFactor;
+    UpdateScale();
+
+    UpdateColors();
+
+    gameObject.SetActive(true);
+  }
+
+  public void Disable() {
+    gameObject.SetActive(false);
+    IsHovered = false;
+    IsDragged = false;
+  }
+
+  private void SetColor(UnityEngine.GameObject go, UnityEngine.Color color) {
+    if (IsInteracting) {
+      UnityEngine.Color.RGBToHSV(color, out float h, out float s, out float v);
+      color = UnityEngine.Color.HSVToRGB(h, s, v * 1.375f, hdr: true);
+    }
+    go.GetComponentInChildren<UnityEngine.Renderer>().material.color = color;
+  }
+
+  private void UpdateColors() {
+    SetColor(base_, XKCDColors.LightGrey);
+    SetColor(tangent_, Style.Tangent);
+    SetColor(normal_, Style.Normal);
+    SetColor(binormal_, Style.Binormal);
+  }
+
+  private void UpdateScale() {
+    hover_scale_offset_ = UnityEngine.Mathf.MoveTowards(hover_scale_offset_,
+      IsInteracting ? 1.25f : 1f,
+      0.075f);
+    transform.localScale =
+        Vector3d.one * normalized_scale_ * hover_scale_offset_;
+  }
+
+  private UnityEngine.Vector3 ScreenManœuvrePosition() =>
+      ScaledToFlattenedScreenPosition(transform.position);
+
+  public void OnMouseEnter() {
+    if (!isActiveAndEnabled) {
+      return;
+    }
+    IsHovered = true;
+  }
+
+  public void OnMouseDown() {
+    mouse_offset_at_click_ =
+        UnityEngine.Input.mousePosition - ScreenManœuvrePosition();
+    IsDragged = true;
+  }
+
+  public void OnMouseDrag() {
+    if (!isActiveAndEnabled) {
+      return;
+    }
+
+    var mouse_offset_now =
+        UnityEngine.Input.mousePosition - ScreenManœuvrePosition();
+    var drag = mouse_offset_now - mouse_offset_at_click_;
+    // Prevent blowup due to wiggly noodles.
+    if (drag.sqrMagnitude < 1f) {
+      return;
+    }
+
+    var screen_tangent =
+        (ScaledToFlattenedScreenPosition(transform.position + tangent_vector_) -
+         ScreenManœuvrePosition()).normalized;
+
+    var projection = UnityEngine.Vector3.Dot(screen_tangent, drag);
+    var Δut = projection *
+              UnityEngine.Mathf.Exp(-mouse_offset_now.sqrMagnitude * 5e-3f);
+    var burn = get_burn_();
+    burn.initial_time += Δut;
+    modify_burn_(burn);
```

Comment:
As mentioned in discussion, this is the part that we should rework to address the stability issues alluded to in the PR description.

My initial thought was that we should be able to minimize the distance to the cursor over the trajectory, and change the time of the burn to the time of that minimum.
Of the local minima found, we need to pick the one closest to the current time of the burn, since there may be multiple revolutions overlapping.

However, we don’t actually have the relevant trajectory: it would be the coast prior to the burn, prolonged past the burn without the burn.

Reply:
FWIW I'm not personally convinced if that level of generality is needed to support the functionality demonstrated above. My thought was that by only allowing interactivity with a preexisting marker, we can get away _without_ searching the entire trajectory, since the problem is reduced to movingly continuously along a continuous curve. Considering the functionality of the kludge implemented above, I'm fairly certain that considering the velocity in line 128 above would result in a nearly indistinguishable user experience compared to full optimization, while being much, much simpler to implement and execute.

But then, Principia isn't exactly known for taking the easy way out, so what am I saying 😆 With optimization, I'd imagine it becomes possible to support hovering along the entirety of the trajectory, stock-style.

Reply:
> we can get away without searching the entire trajectory, since the problem is reduced to movingly continuously along a continuous curve.

The continuity is an interesting point. What really worries me here is the kludge at ll 118–121, but perhaps we can get rid of it while still benefiting from continuity.

> With optimization, I'd imagine it becomes possible to support hovering along the entirety of the trajectory, stock-style.

This would be a different interaction which would require different UI wizardry; but searching along the flight plan for the nearest point to _create_ a manœuvre would in a sense be easier, because then we can just look at the real flight plan trajectory.

Reply:
> What really worries me here is the kludge at ll. 118–121

TBH I'm not sure if it's necessary. I added it while debugging another issue (due to bad math). In fact, I'm not sure if either of the manipulations I'm doing here (that kludge and the exponential decay) are necessary.